sstables/storage.cc

```diff
@@ -36,10 +36,43 @@
 
 namespace sstables {
 
+class opened_directory final {
+    std::filesystem::path _pathname;
+    file _file;
+
+public:
+    explicit opened_directory(std::filesystem::path pathname) : _pathname(std::move(pathname)) {};
+    explicit opened_directory(const sstring &dir) : _pathname(std::string_view(dir)) {};
+    opened_directory(const opened_directory&) = delete;
+    opened_directory& operator=(const opened_directory&) = delete;
+    opened_directory(opened_directory&&) = default;
+    opened_directory& operator=(opened_directory&&) = default;
+    ~opened_directory() = default;
+
+    const std::filesystem::path::string_type& native() const noexcept {
+        return _pathname.native();
+    }
+
+    const std::filesystem::path& path() const noexcept {
+        return _pathname;
+    }
+
+    future<> sync(io_error_handler error_handler) {
+        if (!_file) {
+            _file = co_await do_io_check(error_handler, open_directory, _pathname.native());
+        }
+        co_await do_io_check(error_handler, std::mem_fn(&file::flush), _file);
+    };
+
+    future<> close() {
+        return _file ? _file.close() : make_ready_future<>();
+    }
+};
```

Comment:
Looks much better now, thanks.

I wonder if we could use also an `open(io_error_handle)` method for first time opening.
It would be better to hit an early failure to open before `sync()` time...

Reply:
@bhalevy 

> I wonder if we could use also an `open(io_error_handle)` method for first time opening.
> It would be better to hit an early failure to open before `sync()` time...

To the extent that I understand this, it seems superfluous / uncalled for. AIUI, we have no need for opening the directory _per se_ through this class; we only open the directory for the sake of flushing (the latter is what ensures that the data hits the media permanently, or reports an error).

Put differently, I'm unaware of any code locations where it would be justified to insert a separate error-checked open action, such that it would not be *immediately* followed by a sync anyway. What we actually want to do is sync (at strategic points); open is just a necessary cost, which is why we're "tucking it away" (caching it).

Reply:
The other side of that coin, is that if we error-check every I/O independently we don't have to keep track of whether it's adjacent to another I/O that would "collect" its error result.

I agree that O_CREAT-less open()s in general don't really need an error check. Usually they're not even real I/O. But it will be less confusing overall if we have a single path.